### DISCUSSION TITLE
Extend robot library

### DISCUSSION THREAD

**Discussion by VitaliiSadovshikov** [visose/Robots#140] [Q&A] [answered]
Hi there, we have some robot cells in [our university](https://minobrnauki.gov.ru/press-center/news/?ELEMENT_ID=42408) and I'd like to add them to the "Robotic" library. Is it possible and is there a guide on how to do it? I've noticed there are two files .xml and .3dm. Even if I understand how to create a .xml file, it is obscure how should look a .3dm (should be there layer, mesh names, etc)? Is there a template?

**Comment by visose** [comment, marked as answer]
Hi, there is some info here: https://github.com/visose/Robots/wiki/Robot-libraries

Regarding the .3dm, you can download a few from https://github.com/visose/Robots/tree/libraries and check how they are structured.

A quick summary is that you should create a parent layer for each mechanism that is empty, but contain nested layers, one for each joint of the mechanism. The parent layer should be named "MechanismType.Manufacturer.Model" and the child layers should be named only as whole numbers starting from 0, where 0 is the base, 1 is the first joint and so on. Each child layer should only contain a single mesh object. The mesh name is not used, only the layer name. There should be no other geometry in the file besides these mesh objects.

Note that a mechanism is not a robot cell, each robot cell is made out of one or more mechanisms. The 3dm file doesn't know about robot cells.

You can also add tools, by creating a single layer named "Tool.Name" and placing the mesh directly in this layer.

**Reply by VitaliiSadovshikov** [reply]
Okay, should try, Much obliged!

**Reply by VitaliiSadovshikov** [reply]
Can you clarify some questions where I have stuck:
1) What does mean the "a" and "b" variables of the Revolute (can't find relevant value in the documentation of the robot)? 
2) I have noticed that the payload value is designated in the .xml file as a single value but actually, it is happening according to the [Payload diagram](https://www.kuka.com/-/media/kuka-downloads/imported/6b77eecacfe542d3b736af377562ecaa/0000205456_en.pdf), is it possible to take it into consideration?
3) How does the plugin recognize where are the joint points (axes) of the robot to move the robot parts (the meshes which are placed in the child layers), according to examples the origins of the meshes don't locate at the joint axes and XML doesn't require 3d coordinates of that axes. I am a little bit confused...

**Reply by visose** [reply]
1. You mean the a and d parameters right? You can find more info in other posts, but in summary, they are distances from one joint to the next. Its a convention to describe the relative position between joints of a mechanism. You can compare the values from a robot in an xml file to the dimensions shown in a diagram like the one you linked above. You will see which dimensions to put where.

2. Robots usually have a "rated payload". This is only used in the plugin to warn the user that the tool is exceeding this value, but it doesn't change the simulation or generated program in any way. It's up to the user if he wants to exceed this weight. The closer you are to the limit, the worse the robot might perform, specially if you don't set up correctly the tool weight and center of mass. If you exceed it too much the controller won't run the program. You can use software from the robot manufacturer to check this as well, that takes into account the tool's center of mass and inertia.

3. The parameters mentioned in the first point are used for this.

**Reply by VitaliiSadovshikov** [reply]
> 1. You mean the a and d parameters right?

Yes, thank you for your help!
